qemu/tests/block_resize.py

```diff
@@ -6,6 +6,165 @@
+                     current_size, expected_size)
+        return True
+    else:
+        test.fail("Block size get from guest is not same as expected.\n"
+                  "Reported: %s\nExpect: %s\n" % (current_size, expected_size))
+    return
+
+
+def get_drive_path_linux(test, session, params, data_image):
+    """
+    Get the drive id in linux guest.
+
+    param data_image: data image name, like image1 or stg.
+    param return: return drive path, like /dev/sda or /dev/sdb.
+    """
+    pattern = r"(serial|wwn)=(\w+)"
+    match = re.search(pattern, params["blk_extra_params_%s" % data_image], re.M)
+    if match:
+        drive_id = match.group(2)
+    else:
+        test.fail("No available tag to get drive id")
+    drive_path = utils_misc.get_linux_drive_path(session, drive_id)
+    if not drive_path:
+        test.error("Failed to get '%s' drive path" % data_image)
+    return drive_path
+
+
+def block_resize_enlarge(vm, session, params, device, block_size):
+    """
+    Enlarge the disk image of device to expected block_size.
+
+    param device: enlarge device.
+    param block_size: enlarge device to block_size.
+    """
+    logging.info("Enlarge disk size to %s in monitor"
+                 % block_size, logging.info)
+    vm.monitor.block_resize(device, block_size)
+    if params.get("os_type") == "windows":
+        drive.extend_shrink_disk(session, params["disk_index"],
+                                 params["disk_letter"],
+                                 operation="extend")
+
+
+def block_resize_shrink(vm, session, params, device, old_disk_size, shrunk_size):
+    """
+    Shrink the disk image of device to expected block_size.
+
+    param device: shrink device.
+    param old_disk_size: the size before shrink.
+    param shrunk_size: want to shrink size, after shrunk, the disk size
+    should equal to "old_disk_size - shrunk_size".
+
+    param return: return the block_size: "old_disk_size-shrunk_size".
+    """
+    if params.get("os_type") == "windows":
+        shrunk_size = drive.extend_shrink_disk(session,
+                                               params["disk_index"],
+                                               params["disk_letter"],
+                                               operation="shrink")
+        shrunk_size = float(utils_misc.normalize_data_size("%sM" % shrunk_size,
+                                                           order_magnitude="B"))
+    block_size = old_disk_size - int(shrunk_size)
+    logging.info("Shrink disk size to %s in monitor"
+                 % block_size, logging.info)
+    vm.monitor.block_resize(device, block_size)
+    return block_size
+
+
+def refresh_disk_after_resize(vm, session, params):
+    """
+    Rescan disk in windows or linux guest.
+    In linux guest, after block_resize:
+    need to rescan for virtio-scsi disk(no need reboot),
+    need to reboot for virtio-blk disk(no need rescan).
```

Comment:
It is not    identical with cfg.

 ide:
        # TODO this is bug that ide not support qcow2 format block resize
        # recently. Will remove this line when the bug fixed.
        no qcow2
        need_reboot = yes
    virtio_blk:
        need_reboot = no
        driver_name = viostor
    virtio_scsi:
        need_reboot = no